File: 백준/Gold/16928. 뱀과 사다리 게임/뱀과 사다리 게임.py
import sys
from collections import deque

input = sys.stdin.readline


def bfs():
    queue = deque()
    queue.append(1)

    visited[1] = True

    while queue:
        x = queue.popleft()

        if x == 100:
            return

        for dice in range(1, 7):
            nx = x + dice

            if nx > 100:
                continue

            if visited[nx]:
                continue

            if nx in ladder:
                nx = ladder[nx]

            if nx in snake:
                nx = snake[nx]

            if not visited[nx]:
                visited[nx] = True
                board[nx] = board[x] + 1
                queue.append(nx)


n, m = map(int, input().split())
ladder = {}
snake = {}
board = [0 for _ in range(101)]
visited = [False for _ in range(101)]

for _ in range(n):
    x, y = map(int, input().split())
    ladder[x] = y

for _ in range(m):
    u, v = map(int, input().split())
    snake[u] = v

bfs()

print(board[100])

File: 백준/Gold/16928. 뱀과 사다리 게임/뱀과 사다리 게임.cc
#include <iostream>
#include <algorithm>
#include <vector>
#include <queue>
#include <list>
#include <string>
#include <hash_map>
#include <unordered_map>
using namespace std;

#define endl '\n'

int n, m;
unordered_map<int, int> ladder;
unordered_map<int, int> snake;
int board[101];
bool visited[101];

void bfs() {
    queue<int> q;
    q.push(1);
    visited[1] = true;

    while (!q.empty()) {
        int x = q.front();
        q.pop();

        if (x == 100) {
            return;
        }

        for (int dice = 1; dice < 7; ++dice) {
            int nx = x + dice;

            if (nx > 100) {
                continue;
            }

            if (visited[nx]) {
                continue;
            }

            if (ladder.find(nx) != ladder.end()) {
                nx = ladder[nx];
            }

            if (snake.find(nx) != snake.end()) {
                nx = snake[nx];
            }

            if (!visited[nx]) {
                visited[nx] = true;
                board[nx] = board[x] + 1;
                q.push(nx);
            }

        }
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);

    cin >> n >> m;

    for (int i = 0; i < n; ++i) {
        int x, y;
        cin >> x >> y;
        ladder.insert(make_pair(x, y));
    }

    for (int i = 0; i < m; ++i) {
        int u, v;
        cin >> u >> v;
        snake.insert(make_pair(u, v));
    }

    bfs();

    cout << board[100];

}
